1-pattern-history-consumer.md

```diff
@@ -0,0 +1,121 @@
+# A pattern for history consumer in event sourcing
+
+Event sourcing is a popular [technique to capture changes to data as a sequence of events](ref1).
+
+Applications interested in this data usually read (consume) the events in order and build an
+accumulated representation often referred to as "the state".
+
+It is also quite common for these applications to build the state in-memory, instead of storing it in
+other slower and persistent mediums.
+In this case, every time the application starts it needs to go through all the events in the log and process them.
+
+If the amount of events is low enough and the processing time is quick enough, then this can lead
+to particularly lightweight micro/nano-services with no permanent storage dependencies.
+The caveat is that the service probably should hold off any work until this state is rebuilt, otherwise
+it might lack information it needs to correctly do its function.
+For example, it might publish events that are already in the log, but since it has not _caught up_
+with it entirely, it publish them again, leading to duplicates, corrupting the log, etc.
+
+In a single process or single threaded system this is a trivial problem to solve.
+Simply ensure that the application first consumes all of the log, and only after that it can start doing its thing.
+
+But that is not how modern applications are built.
+Even if it takes just a few seconds to process all of the log,
+users expect the application to be responsive during that time.
+For example, service health and _readyness_ checks should respond accordingly.
```

Comment:
Worth a mention to something like an API used by a frontend as an example too?

Reply:
I like that!